Feature/40 record revisions metadata

pospi opened the pull request:
Made a first pass at metadata conversion helpers and am curious to get an early review from someone who has been in the weeds with the DHT before to see if there are any glaring mistakes.

Would also really appreciate pointers or stubbing out of how to test this with the HDK mocking framework as I haven't done that before and it would take me much longer to find my feet with it (:

pospi committed:
implement first pass of metadata helpers

loads previous revision & future revisions metadata from DHT Element

minor edits to record_helpers for code reuse

pospi committed:
stub out record conflict error type

pospi committed:
implement first pass at first revision determination & previous revisions counting

Connoropolous commented on lib/hdk_records/src/metadata_helpers.rs:
```diff
@@ -0,0 +1,186 @@
+pub struct RevisionMeta {
+    pub id: HeaderHash,
+    pub time: Timestamp,
+    pub author_id: AgentPubKey,
```
is author_id as a property name part of the Valueflows spec?

pospi replied:
It is, but I take your point- there's no reason the two have to marry and actually this one doesn't anyway (it will have to be translated through a `whois` query against the REA Agent DNA in the GraphQL resolvers in order to be turned into an `AgentId` once #219 has been implemented). What would make more sense as idiomatically Holochain? `agent_pubkey`?

Connoropolous replied:
Yeah something like... agent_pub_key to match the casing 👍

Connoropolous reviewed (commented):
the pattern looks broadly sensible to me. I can do something with the MockHDK soon here, yeah

Connoropolous commented:
so @pospi you mean starting in on #214 would be helpful here?

pospi committed:
Merge branch 'sprout' into feature/40-record-revisions-metadata

Commit:
Merge branch 'sprout' into feature/40-record-revisions-metadata

pospi committed:
Merge branch 'sprout' into feature/40-record-revisions-metadata

pospi committed:
update core record management lib to pass through metadata struct rather than raw HeaderHash

pospi committed:
export RevisionMeta from VF attributes crate for use in RPC structs

does imply dependency on HDK and break purity of RPC crates, but not worth optimising until thirdparty Rust apps need this layer of integration

pospi committed:
update lockfile

pospi committed:
update all externally facing CRUD API endpoints to include author metadata in response payloads

pospi committed:
Merge branch 'feature/40-record-revisions-metadata' of github.com:h-REA/hREA into feature/40-record-revisions-metadata

pospi committed:
rename metadata field to be more Holochain idiomatic

pospi commented:
@Connoropolous have pushed a first run at this. If CI passes, we should merge. Otherwise we should prioritise finishing off before continuing on #340 as the changes are quite extensive and I expect there will be significant merge conflicts if we attempt both at once.

pospi commented:
Still to do: needs some tests in `core-architecture` to assert that the data is returned and that it changes after an update. There may also be some work to do in the GraphQL layer but I'm pretty sure both `id` and `time` will be passed through without intervention. Possibly the `history` module needs to be enabled in the VF-GraphQL modules whitelist as well.

Anyway, nearly there I think! Signing off for the weekend (:

Connoropolous commented:
nice! yes I see CI passing and I did a basic review of the code and saw no issues
